octodns_cloudflare/__init__.py

```diff
@@ -1209,23 +1227,77 @@ def _apply_Delete(self, change):
                     )
                     self._try_request('DELETE', path)
 
+    def _supported_plans(self, zone_name):
+        zone_id = self.zones[zone_name]['id']
+        path = f'/zones/{zone_id}/available_plans'
+        resp = self._try_request('GET', path)
+        result = resp['result']
+        if isinstance(result, list):
+            return [plan['legacy_id'] for plan in result]
+        msg = f'{self.id}: unable to determine supported plans, do you have an Enterprise account?'
+        raise SupportsException(msg)
+
+    def _update_plan(self, zone_name, plan):
+        current_plan = self.zones[zone_name].get('plan', False)
+        if current_plan and current_plan == plan:
+            return
+        if plan in self._supported_plans(zone_name):
+            zone_id = self.zones[zone_name]['id']
+            data = {'plan': {'legacy_id': plan}}
+            resp = self._try_request('PATCH', f'/zones/{zone_id}', data=data)
+            # Update the cached plan information
+            self.zones[zone_name]['plan'] = resp['result']['plan']['legacy_id']
+        else:
+            msg = f'{self.id}: {plan} is not supported for {zone_name}'
+            raise SupportsException(msg)
+
+    def _plan_meta(self, existing, desired, changes):
+        meta = {}
+        zone_name = desired.name
+        desired_plan = self.plan_type
+        if zone_name in self.zones:
+            current_plan = self.zones[zone_name].get('plan', False)
+            if desired_plan is None and current_plan is not False:
+                # If desired plan is removed and current plan is set, set to free plan
+                meta['cloudflare_plan'] = CLOUDFLARE_FREE_PLAN
+            elif current_plan != desired_plan:
+                meta['cloudflare_plan'] = desired_plan
+        return meta
```

Comment:
With avoiding the supported plans lookup above and leaving things unmanaged/as-is when the provider doesn't have a value I believe this section can be removed and simplified to:

```suggestion
            if current_plan != desired_plan:
                return {
                    'cloudflare_plan': {
                        'current': current_plan,
                        'desired': desired_plan,
                    }
        return
```

I also went ahead and put a bit more info into the meta saying what the current value is and what it will be to make it more similar to the other information in a plan. Seems like a reasonable best practice for it. I should probably update the doc on the base method to say so if we go that route here.

One part I'm not sure about is the default value of `False` up above causing a persistent change if for some reason the zones list doesn't ever have a value when it's being built.

Reply:
The logic here was to support removing a zone: Cloudflare requires Enterprise plan customers to set the plan first to `free` for some reason (https://developers.cloudflare.com/fundamentals/setup/manage-domains/remove-domain/). Perhaps it is better to leave that out from here 👍